test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,552 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+use serde_json::Value;
+use serde::{Deserialize, Deserializer};
+use up_rust::{
+    Data, UAttributes, UAuthority, UCode, UEntity, UMessage, UMessageType, UPayload,
+    UPayloadFormat, UPriority, UResource, UUri, UUID,
+};
+
+use protobuf::{MessageField, SpecialFields};
+
+pub fn convert_json_to_jsonstring<T: serde::Serialize>(value: &T) -> String {
+    serde_json::to_string(value).expect("Failed to convert to JSON string")
+}
+
+#[derive(Debug, Default)]
+pub struct WrapperUUri(pub UUri);
+impl<'de> Deserialize<'de> for WrapperUUri {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
+        //dbg!("WrapperUUri: {:?}", value.);
+
+        //update authority
+        let _authority_name = match value
+            .get("authority")
+            .and_then(|authority| authority.get("name"))
+        {
+            Some(_authority_name) => _authority_name.as_str(),
+            None => Some("default"),
+        };
+
+        let _authority_number_ip = match value
+            .get("authority")
+            .and_then(|authority| authority.get("number"))
+            .and_then(|number| number.get("Ip"))
+        {
+            Some(_authority_number_ip) => _authority_number_ip.to_string().as_bytes().to_vec(),
+            None => {
+                let default: Vec<u8> = vec![0];
+                default
+            }
+        };
+        let _authority_number_id = match value
+            .get("authority")
+            .and_then(|authority| authority.get("number"))
+            .and_then(|number| number.get("Id"))
+        {
+            Some(_authority_number_id) => _authority_number_id.to_string().as_bytes().to_vec(),
+            None => {
+                let default: Vec<u8> = vec![0];
+                default
+            }
+        };
+
+        let mut _authority = UAuthority::new();
+        if!( _authority_name.clone() == Some("default")){
+            _authority.name = _authority_name.map(|s| s.to_string());
+        }
+        if!(_authority_number_id.clone() == vec![0]){
+            _authority.set_id(_authority_number_id.clone());
+        }
+        if!(_authority_number_ip.clone() == vec![0]){
+            _authority.set_ip(_authority_number_ip.clone());
+        }
+
+        //update entity
+        let _entity_name = match value.get("entity").and_then(|entity| entity.get("name")) {
```

Comment:
Same sorts of suggestions as above:
* try to chain together to get what you need
* try to model these chains as resulting in an `Option<T>` so that later you can treat them ergonomically
* create the `UEntity` at the very end so that you can avoid mutation